lang/shared/src/main/scala/com/wavesplatform/lang/v1/testing/ScriptGen.scala

```diff
@@ -73,14 +73,28 @@ trait ScriptGen {
       value <- BOOLgen((gas - 3) / 3)
     } yield LET(name, value)
 
+  def REFgen: Gen[EXPR] =
+    Gen.identifier.map(REF)
+
   def BLOCKgen(gas: Int): Gen[EXPR] =
     for {
-      let  <- LETgen((gas - 3) / 3) // should be Gen.option(LETgen((gas - 3) / 3)), issue: NODE-696
-      body <- BOOLgen((gas - 3) / 3)
+      let  <- LETgen((gas - 3) / 3)                                      // should be Gen.option(LETgen((gas - 3) / 3)), issue: NODE-696
+      body <- Gen.oneOf(BOOLgen((gas - 3) / 3), BLOCKgen((gas - 3) / 3)) // BLOCKGen wasn't add to BOOLGen since issue: NODE-700
+      // ParseError example:
+      //(let xtld = true; true && (true&&true))
     } yield BLOCK(Some(let), body)
 
   private val spaceChars: Seq[Char] = Vector('\u0020', '\u0009', '\u000D', '\u000A')
 
   def whitespaceChar: Gen[Char] = Gen.oneOf(spaceChars)
   val whitespaces: Gen[String]  = Gen.listOf(whitespaceChar).map(_.mkString)
 }
+
+trait ScriptGenParser extends ScriptGen {
+  override def BOOLgen(gas: Int): Gen[EXPR] = {
```

Comment:
This is not needed. `ScriptGen` is not a stateful a stateful object, so no inheritance-oop is neccessary. 
I suggest the following: Ensure that you really need 2 implementations of `IntGen` if so(and that's ok!), just parameterise `BLOCKgen` so that it'd accept `INTGen` and `BOOLgen`.

Reply:
yep, we need two implementations of INTgen and BOOLgen because of difference with context usage for parser and typechecker. For one we don't need references, but for another it is necessary to have valid reference to param.